src/plugins/clips-executive/clips/domain.clp

```diff
@@ -596,6 +597,11 @@
         (goal-id ?g) (plan-id ?p)
         (part-of ?pn) (grounded TRUE)
         (grounded-with ?action-id) (is-satisfied FALSE)))
+
+  ; ensure that we have grounded all related atomic preconditions before
+  (forall (domain-atomic-precondition (operator ?op) (part-of ?pn) (name ?apname) (grounded FALSE))
+    (domain-atomic-precondition (operator ?op) (part-of ?pn) (name ?apname) (grounded TRUE))
+  )
```

Comment:
Aren't these contradicting? In my understanding, this condition checks if for each ungrounded domain-atomic-precondition, there is a grounded one with the same parameters. I can't say that I have a deep understanding about the precondition checking and action grounding, but this seems odd to me.

Reply:
What we do here is the following:
for all atomic preconditions which are not grounded (and in extension not bound to an action, thus being the general precondition description of actions), there must also be a grounded one that is linked to the specific action instance (thus, the grounding for all atomic preconditions associated to the compound precondition has happened).

In a formula it could be "forall x. exists y. p(x) /\ !g(x) => p(y) /\ g(y)" where p denotes precondition atoms and g groundedness.

Reply:
To put it differently: When we ground the precondition of an operator, we copy all the sub-conditions, so the ungrounded precondition is unmodified. That's why we can use the ungrounded precondition to check if all sub-conditions have been grounded.